metapp/metapp.ml

```diff
@@ -371,9 +371,30 @@ class filter =
     check_expr (snd arg) in
   let check_type_declaration (declaration : Ppxlib.type_declaration) =
     check_attr declaration.ptype_attributes in
+  let check_constructor (c : Ppxlib.constructor_declaration) =
+    check_attr c.pcd_attributes
+  in
+  let check_label (l: Ppxlib.label_declaration) =
+    check_attr l.pld_attributes
+  in
   object
     inherit Ppxlib.Ast_traverse.map as super
 
+    method! type_declaration (t : Ppxlib.type_declaration) : Ppxlib.type_declaration =
+      let t = super#type_declaration t in
+      match t.ptype_kind with
+      | Ptype_variant cons ->
+        begin  match List.filter check_constructor cons with
+        | [] ->{ t with ptype_kind = Ptype_abstract }
+        | cons -> { t with ptype_kind = Ptype_variant cons }
+        end
+      | Ptype_record labels ->
+        begin  match List.filter check_label labels with
+        | [] ->{ t with ptype_kind = Ptype_abstract }
+        | labels -> { t with ptype_kind = Ptype_record labels }
+        end
```

Comment:
Empty variant types are allowed since OCaml 4.07.0 (https://v2.ocaml.org/manual/emptyvariants.html). Empty records are not allowed, but I would prefer an error in this case.

```
begin  match List.filter check_label labels with
        | [] ->Location.raise_errorf ~loc:t.ptype_loc "Empty records are not allowed"
```

(Otherwise, we could argue that an empty record should be a type alias for `unit`, but I think it is better to ask the user to be explicit about what she wants.]